examples/4_totp_storage.py

```diff
@@ -0,0 +1,218 @@
+###################################
+
+import asyncio
+import json
+import logging
+import os
+import sys
+from enum import Enum
+
+import pyotp
+
+from pykeybasebot import Bot, EventType
+
+logging.basicConfig(level=logging.DEBUG)
+
+if "win32" in sys.platform:
+    # Windows specific event-loop policy
+    asyncio.set_event_loop_policy(asyncio.WindowsProactorEventLoopPolicy())
+
+
+class TotpMsg(Enum):
+    PROVISION = "provision"
+    REPROVISION = "reprovision"
+    REMOVE = "remove"
+    NOW = "now"
+    URI = "uri"
+    LIST = "list"
+
+
+class TotpHandler:
+    """
+    TotpHandler handles commands sent via chat and uses the team key-value store to
+    provide TOTP client functionality.
+
+    TotpHandler listens to chat messages of the form:
+
+    `!totp {provision|reprovision|remove|now|uri} <issuer>` and
+    `!totp list`
+
+    For each provisioned key, the handler stores in the namespace "totp" one
+    row, with the key "<issuer" and the json blob value "{"secret": <16 char base32
+    secret>}".
+
+    This Keybase bot can be used in place of MFA apps like Google Authenticator and Authy
+    for login processes that require two-step verification. This bot uses the
+    pyotp library, so it is also compatible with other MFA apps.
+
+    A Keybase TOTP bot could be useful for logins where you're already using a strong
+    password, but are required to also use TOTP. Or it could be perfect for adding
+    replay attack resistance to the physical keypad lock protecting your team's ice
+    cream freezer.... Note that depending on the login process and how you use this bot
+    with your Keybase team setup, this bot may not be appropriate for threat models
+    that require a physical "second factor" device.
+
+    For more information on TOTP, see https://pyotp.readthedocs.io/en/latest/,
+    https://tools.ietf.org/html/rfc6238.
+    """
+
+    MSG_PREFIX = "!totp"
+    NAMESPACE = "totp"
+
+    def to_json(self, secret):
+        return {"secret": secret}
+
+    async def __provision(self, bot, team, issuer, force: bool = False):
+        secret = pyotp.random_base32()
+        val = json.dumps(self.to_json(secret))
+        if not force:
+            try:
+                # throws exception if 1 is not the latest revision + 1
+                await bot.kvstore.put(team, self.NAMESPACE, issuer, val, revision=1)
+            except Exception as e:
+                res = await bot.kvstore.get(team, self.NAMESPACE, issuer)
+                if bot.kvstore.is_deleted(res):
+                    # then TOTP for this issuer was previously provisioned but
+                    # was deleted. insert as latest revision + 1
+                    await bot.kvstore.put(team, self.NAMESPACE, issuer)
+                else:
+                    # then TOTP for this issuer was previously provisioned and
+                    # might still be in use, so raise the exception
+                    raise e
```

Comment:
in this case, if it's actually a revision error, we can definitely convert that from kvstore-revision-wording to totp-contextually-meaningful-something. like "it looks like you already provisioned this. try deleting it?" <-- or whatever makes sense.